vhost-device-template/src/vhu_foo.rs

```diff
@@ -0,0 +1,394 @@
+    }
+}
+
+pub(crate) struct VhostUserFooBackend {
+    info: FooInfo,
+    event_idx: bool,
+    pub exit_event: EventFd,
+    mem: Option<GuestMemoryLoadGuard<GuestMemoryMmap>>,
+}
+
+type FooDescriptorChain = DescriptorChain<GuestMemoryLoadGuard<GuestMemoryMmap<()>>>;
+
+impl VhostUserFooBackend {
+    pub fn new(info: FooInfo) -> Result<Self> {
+        Ok(VhostUserFooBackend {
+            info,
+            event_idx: false,
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
+            mem: None,
+        })
+    }
+
+    /// Process the requests in the vring and dispatch replies
+    fn process_requests(
+        &mut self,
+        requests: Vec<FooDescriptorChain>,
+        _vring: &VringRwLock,
+    ) -> Result<()> {
+        if requests.is_empty() {
+            println!("No pending requests\n");
+            return Ok(());
+        }
+
+        // Iterate over each FOO request.
+        for desc_chain in requests.clone() {
+            let counter = self.info.counter();
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            println!(
+                "Request number: {} contains {} descriptors\n",
+                counter,
+                descriptors.len(),
+            );
+
+            for (i, desc) in descriptors.iter().enumerate() {
+                // One can also read/write the contents of the descriptors here.
+                println!("Length of descriptor@{} is: {}\n", i, desc.len());
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Process the requests in the vring and dispatch replies
+    fn process_queue(&mut self, vring: &VringRwLock) -> Result<()> {
+        // Collect all pending requests
+        let requests: Vec<_> = vring
+            .get_mut()
+            .get_queue_mut()
+            .iter(self.mem.as_ref().unwrap().clone())
+            .map_err(|_| Error::DescriptorNotFound)?
+            .collect();
+
+        if self.process_requests(requests, vring).is_ok() {
+            // Send notification once all the requests are processed
+            vring
+                .signal_used_queue()
+                .map_err(|_| Error::NotificationFailed)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// VhostUserBackendMut trait methods
+impl VhostUserBackendMut for VhostUserFooBackend {
+    type Vring = VringRwLock;
+    type Bitmap = ();
+
+    fn num_queues(&self) -> usize {
+        NUM_QUEUES
+    }
+
+    fn max_queue_size(&self) -> usize {
+        QUEUE_SIZE
+    }
+
+    fn features(&self) -> u64 {
+        // this matches the current libvhost defaults except VHOST_F_LOG_ALL
+        1 << VIRTIO_F_VERSION_1
+            | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
+            | 1 << VIRTIO_RING_F_INDIRECT_DESC
+            | 1 << VIRTIO_RING_F_EVENT_IDX
+            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
+    }
+
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
+        VhostUserProtocolFeatures::MQ
```

Comment:
Is MQ always a feature we want? I think we should point out protocol_features doesn't need to be defined unless required and must be accompanied by the supporting PROTOCOL_FEATURES bits in features. If protocol features are enabled where do you fill out your handling of such messages?

Reply:
IMHO if we aim for standalone daemons we should encourage people to advertise CONFIG. MQ is still a bit weird to me since it is a bit ambiguous with CONFIG also providing the same info. But I think vhost-user currently recommends you to implement MQ too.

Reply:
> Is MQ always a feature we want?

https://github.com/qemu/qemu/blob/master/docs/interop/vhost-user.rst#id44

```
Back-ends should always implement the VHOST_USER_PROTOCOL_F_MQ protocol feature, even for devices with a fixed number of virtqueues, since it is simple to implement and offers a degree of introspection.
```

Reply:
Added a comment for protocol features.